meson.build

```diff
@@ -13,256 +104,477 @@ uvsrc = [
+    'test/test-signal-pending-on-close.c',
+    'test/test-signal.c',
+    'test/test-socket-buffer-size.c',
+    'test/test-spawn.c',
+    'test/test-stdio-over-pipes.c',
+    'test/test-strscpy.c',
+    'test/test-tcp-alloc-cb-fail.c',
+    'test/test-tcp-bind-error.c',
+    'test/test-tcp-bind6-error.c',
+    'test/test-tcp-close-accept.c',
+    'test/test-tcp-close-while-connecting.c',
+    'test/test-tcp-close.c',
+    'test/test-tcp-close-reset.c',
+    'test/test-tcp-connect-error-after-write.c',
+    'test/test-tcp-connect-error.c',
+    'test/test-tcp-connect-timeout.c',
+    'test/test-tcp-connect6-error.c',
+    'test/test-tcp-create-socket-early.c',
+    'test/test-tcp-flags.c',
+    'test/test-tcp-oob.c',
+    'test/test-tcp-open.c',
+    'test/test-tcp-read-stop.c',
+    'test/test-tcp-read-stop-start.c',
+    'test/test-tcp-shutdown-after-write.c',
+    'test/test-tcp-try-write.c',
+    'test/test-tcp-try-write-error.c',
+    'test/test-tcp-unexpected-read.c',
+    'test/test-tcp-write-after-connect.c',
+    'test/test-tcp-write-fail.c',
+    'test/test-tcp-write-queue-order.c',
+    'test/test-tcp-write-to-half-open-connection.c',
+    'test/test-tcp-writealot.c',
+    'test/test-test-macros.c',
+    'test/test-thread-equal.c',
+    'test/test-thread.c',
+    'test/test-threadpool-cancel.c',
+    'test/test-threadpool.c',
+    'test/test-timer-again.c',
+    'test/test-timer-from-check.c',
+    'test/test-timer.c',
+    'test/test-tmpdir.c',
+    'test/test-tty-duplicate-key.c',
+    'test/test-tty-escape-sequence-processing.c',
+    'test/test-tty.c',
+    'test/test-udp-alloc-cb-fail.c',
+    'test/test-udp-bind.c',
+    'test/test-udp-connect.c',
+    'test/test-udp-create-socket-early.c',
+    'test/test-udp-dgram-too-big.c',
+    'test/test-udp-ipv6.c',
+    'test/test-udp-mmsg.c',
+    'test/test-udp-multicast-interface.c',
+    'test/test-udp-multicast-interface6.c',
+    'test/test-udp-multicast-join.c',
+    'test/test-udp-multicast-join6.c',
+    'test/test-udp-multicast-ttl.c',
+    'test/test-udp-open.c',
+    'test/test-udp-options.c',
+    'test/test-udp-send-and-recv.c',
+    'test/test-udp-send-hang-loop.c',
+    'test/test-udp-send-immediate.c',
+    'test/test-udp-sendmmsg-error.c',
+    'test/test-udp-send-unreachable.c',
+    'test/test-udp-try-send.c',
+    'test/test-uname.c',
+    'test/test-walk-handles.c',
+    'test/test-watcher-cross-stop.c',
+  ]
+  if win32
+    # TODO: not sure if this works like that, it doesn't on linux
+    libuv_test_src += meson.source_root() + '/uv_win_longpath.manifest'
+  endif
+  libuv_test_exe = executable('uv_run_tests', libuv_test_src,
+    dependencies: [libuv_dep, libuv_test_deps], c_args: libuv_cargs)
+  test('libuv_run_tests', libuv_test_exe, workdir: meson.current_source_dir(),
+    protocol: 'tap', timeout: 600, is_parallel: false)
+
+  libuv_test_a_lflags = libuv_qemu ? '-static' : []
+  libuv_test_a_exe = executable('uv_run_tests_a', libuv_test_src,
```

Comment:
As a matter of curiosity -- why explicitly build both static and shared libs, and test both of them? Shouldn't people be using default_library to select one or the other?

Reply:
I basically followed the original cmake, because from my point of view, both variants aren't using the exact same code, because the shared variant sets two defines: USING_UV_SHARED and BUILDING_UV_SHARED.

I just checked and it seems to be because of win32-specific stuff (`__declspec(dllexport), __declspec(dllimport)`).

Reply:
That's unfortunately a well-known problem with Microsoft. It might be only needed by msvc, not by the MinGW gcc.

There is an open issue for meson to support setting different c_args on the static/shared parts of a library()/both_libraries() but no merged code.

Another solution is to get_option('default_library'), choose which defines to use... and if the value is "both", just raise error('default_library=both is not supported on Windows').

Reply:
did that now, lgtm tbh